src/app/cli/src/commands/ingest_command.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use kamu::domain::*;
+use opendatafabric::*;
+
+use super::{CLIError, Command};
+use crate::OutputConfig;
+
+///////////////////////////////////////////////////////////////////////////////
+// Command
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct IngestCommand {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    ingest_svc: Arc<dyn IngestService>,
+    output_config: Arc<OutputConfig>,
+    remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+    dataset_ref: DatasetRef,
+    files_refs: Vec<String>,
+    stdin: bool,
+    recursive: bool,
+}
+
+impl IngestCommand {
+    pub fn new<'s, I>(
+        dataset_repo: Arc<dyn DatasetRepository>,
+        ingest_svc: Arc<dyn IngestService>,
+        output_config: Arc<OutputConfig>,
+        remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+        dataset_ref: DatasetRef,
+        files_refs: I,
+        stdin: bool,
+        recursive: bool,
+    ) -> Self
+    where
+        I: Iterator<Item = &'s str>,
+    {
+        Self {
+            dataset_repo,
+            ingest_svc,
+            output_config,
+            remote_alias_reg,
+            dataset_ref,
+            files_refs: files_refs.map(|s| s.to_string()).collect(),
+            stdin,
+            recursive,
+        }
+    }
+
+    fn path_to_url(path: &str) -> Result<url::Url, CLIError> {
+        let p = PathBuf::from(path)
+            .canonicalize()
+            .map_err(|e| CLIError::usage_error(format!("Ivalid path {}: {}", path, e)))?;
+        url::Url::from_file_path(p)
+            .map_err(|_| CLIError::usage_error(format!("Ivalid path {}", path)))
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl Command for IngestCommand {
+    async fn run(&mut self) -> Result<(), CLIError> {
```

Comment:
Minor: the method is quite long. Try splitting out validations and progress tracking inits.

Reply:
Dome on my current branch